lib/internal/source_map/source_map.js

```diff
@@ -205,6 +205,32 @@ class SourceMap {
     };
   }
 
+  /**
+   * @param {number} lineNumber 1-indexed line number in compiled resource call site
+   * @param {number} columnNumber 1-indexed column number in compiled resource call site
+   * @return {object} representing origin call site if found, or empty object
+   */
+  findOrigin(lineNumber, columnNumber) {
+    const range = this.findEntry(lineNumber - 1, columnNumber - 1);
+    if (
+      range.originalSource === undefined ||
+      range.originalLine === undefined ||
+      range.originalColumn === undefined ||
+      range.generatedLine === undefined ||
+      range.generatedColumn === undefined
+    ) {
+      return {};
+    }
+    const lineOffset = lineNumber - range.generatedLine;
+    const columnOffset = columnNumber - range.generatedColumn;
```

Comment:
I tried this PR out locally and noticed that if you create a source map from an existing coverage report some inputs return negative column numbers. Is this function expected to work with the source map data generated via `NODE_V8_COVERAGE`?

Reply:
Oh that's weird. No, the column number should be at minimum 1. Can you share the steps you used to get to a negative column number? Maybe you're passing it offsets instead of column numbers?

Reply:
@isaacs - I'll have to see if I still have the code on a branch, but the gist of it was:

Given a line in the generated file, I was trying to find all of the corresponding lines in the original source file. I created a `SourceMap` from the V8 coverage data and called `map.findOrigin(row, 1)` and `map.findOrigin(row, Infinity)`. Note - I'm not sure if `Infinity` should be supported here, but I did try valid finite numbers. Some of the returned values had negative column offsets. Using zero-based rows and columns, this did seem to work correctly with `map.findEntry()`, including passing `Infinity`.

Reply:
I would assume that `findOrigin(row, Infinity)` would produce a very strange value indeed, but I'd expect a column of `Infinity` rather than a negative value.

What it's doing is calculating the row and column deltas from the start of the generated line/column offsets to the 1-indexed line/column numbers you give it, then applying that same delta to the source line/column values from the map entry.  So if you told it your origin is `line: 100, column: Infinity` then `findEntry` should give you a `generatedLine` of some number less than 100 (whatever line that range starts on), and a `generatedColumn` of the 0-index start of that range.  The line offset then would be some number less than 100, and the column offset would be `Infinity - (some finite n)`, so Infinity.  The source line and column would be some arbitrary number corresponding to the start of that range, and then it would add the line delta (some number less than 100) to the entry's sourceLine, and the column delta (Infinity) to the entry sourceColumn.

I'm not sure just looking for `line, 0` and `line, Infinity` would give you all ranges that affect that line, though, since more than 2 ranges can exist on a given line, and they might be arbitrarily intermixed.

As I describe this, I'm realizing that this strategy might only work because TypeScript doesn't change token lengths, though.

Like, let's say you had a source like

```js
const someLongName = otherLongName.x()
```

And that got minified to

```js
const _l=_o.x()
```

And let's say the error being thrown is that `otherLongName`/`_o` is undefined, so `_o.x()` throws an error.

Lines are the same, so no issue there.

But the range will start at `{generatedColumn:0,sourceColumn:0}`, and the error origin callsite will be column 13:

```
can't call method x() of undefined blah blah blah
const _l=_o.x()
            ^
```

If we add 13 to the sourceColumn, that isn't going to point at the right spot, and you'll get an output like:

```
const someLongName = otherLongName.x()
            ^
```

It'll work if the minifier is smart enough to make a new range where the two start matching, then it's fine, but I think the heuristic might have to be a little bit smarter, take into account the _end_ of the range as well, and idk, say "It's probably somewhere in the general vicinity of these characters, good luck".

Reply:
Anyway, yes, some source maps and generated/origin files demonstrating the issue would be super handy.  Even if it somehow means this can't live in core, at least I can make tap's stack management thing do the right thing, or close to it.

Reply:
Looking into this a bit more, I think this strategy might actually be fine.

The source maps created by rollup are pretty much perfect. Every token that isn't mangled is source mapped to the appropriate location, and the only things that appear in a range are 1:1 mapping. Uglify's are a bit harder to inspect, because they don't link/include the sources (at least, not in a way that VSCode's visualizer can understand), but it seems like they also map tokens such that offset deltas should be preserved.

Looking at the spec, it seems like this might be required? I'm far from an expert, though.

Reply:
Ah, so the problem is with `findEntry`.  There's no entry in the sourcemap that covers offset `10, 0`.  So, `findEntry` returns the entry at `9,30`, since it uses the highest start of the range that is less than the provided values.

The delta from the range `9,30` to the call site location `11,1` is `+2,-29`.  But really, that's invalid, because `10, 0` isn't covered by the range that starts at offset `9,30`.  There's no way to ever have a call site that references line 11 column 1, and even if it did, that doesn't map to anything in the source file.  It's an invalid input, so you get an invalid result.

Similarly, the range offset `10, Infinity` (incorrectly) returns the range at `10, 1` (because Infinity is greater than the last range in line 9, and 9 is less than the first offset in line 10, so you get that one).  This maps to the source offset of `9,1`.  The delta from `10, 1` to `11, Infinity` is `+1, +Infinity`, so you get a resulting origin call site of `10, Infinity`.

I think that `findEntry` should return an empty object (like it does when it can't find a range) if the provided offsets are outside the end of the offset it finds.  The code seems to assume that there's no un-mapped ranges in the generated file, and that's not the case.

Also, just to un-XY this problem, if your goal is "find all the lines in the origin source that map to a given line in the generated source", this strategy of checking column 1 and column Infinity is not ever going to work reliably. Any line in the generated source may be composed of an arbitrary number of entries from an arbitrary set of locations within the origin source, so if you only test the first and last in the line, you'll miss everything in the middle, and there's no guarantee that they line up. Consider, for example, a munger that shuffles functions around, or a minifier that puts _every_ line in the origin into a single line in the output.

What you need to do is get the set of offset ranges from the sourcemap entries that have a line offset of `line - 1`, and then add `1` to each of the origin lines of those entries, and that'd give you the result.  However, since `sourceMap.#mappings` is private, and only exposed via the very limited `findEntry` method, that's not possible without parsing the source map in userland, unless we also add a way to get at the raw entries themselves.

Reply:
Ie, suggesting this change to findEntry, to have it return an empty range if the supplied offsets are not within the entry it finds (haven't tested, this might break other stuff, idk)

```patch
diff --git a/lib/internal/source_map/source_map.js b/lib/internal/source_map/source_map.js
index 3112a026b64..08ad65b919a 100644
--- a/lib/internal/source_map/source_map.js
+++ b/lib/internal/source_map/source_map.js
@@ -189,10 +189,7 @@ class SourceMap {
       }
     }
     const entry = this.#mappings[first];
-    if (!first && entry && (lineOffset < entry[0] ||
-        (lineOffset === entry[0] && columnOffset < entry[1]))) {
-      return {};
-    } else if (!entry) {
+    if (!entry || (lineOffset !== entry[0] || columnOffset < entry[1])) {
       return {};
     }
     return {
```

With that, it's much more clear that the approach of testing column 1 and column Infinity is invalid, because you walked off the source map.

Reply:
Actually, that's only half of the fix, because we're not tracking the _end_ of the range in `this.#mappings`, only the start.

Reply:
> Also, just to un-XY this problem, if your goal is "find all the lines in the origin source that map to a given line in the generated source", this strategy of checking column 1 and column Infinity is not ever going to work reliably.

So this was just me doing some random testing, not an actual goal. I agree that it won't work for anything useful - a generated file can be a single line, which would then map to every line in the original source, making it pretty useless for something like code coverage.

Reply:
I can definitely see some potential for something like "from `l1, c1` to `l2, c2`, find all the ranges in the origin that are affected", for which "find all the origin lines that went into line X in generated source" is sort of a simplified case. But we can't do that without parsing out the ends of the ranges from the sourcemap payload.